ladybug_geometry/geometry3d/face.py

```diff
@@ -0,0 +1,1239 @@
+        max_subh = 0.98 * parent_height
+        sub_rect_height = max_subh if sub_rect_height > max_subh else sub_rect_height
+        # if sill_height is close to 0, set it to just above 0.
+        min_sill = 0.01 * parent_height
+        sill_height = min_sill if sill_height < min_sill else sill_height
+        # properties used throughout the computation of sub-rectangles
+        bottom_seg = LineSegment3D.from_sdl(base_plane.o, base_plane.x, parent_base)
+
+        if target_area < max_area_subdiv:
+            # divide up the rectangle into points on the bottom.
+            if parent_base > (horizontal_separation / 2):
+                num_div = round(parent_base / horizontal_separation, 0)
+            else:
+                num_div = 1
+            btm_div_pts = bottom_seg.subdivide_evenly(num_div)
+            btm_div_segs = tuple(LineSegment3D.from_end_points(pt, btm_div_pts[i + 1])
+                                 for i, pt in enumerate(btm_div_pts[:-1]))
+            # move the segments to the sill height
+            max_sill_h = parent_height * 0.99 - sub_rect_height
+            sill_vec = base_plane.y * sill_height if sill_height < max_sill_h \
+                else base_plane.y * max_sill_h
+            div_segs = tuple(seg.move(sill_vec) for seg in btm_div_segs)
+            # scale the segments along their center points
+            seg_width = div_segs[0].length
+            subrect_width = (target_area / sub_rect_height) / num_div
+            scale_fac = subrect_width / seg_width
+            scaled_segs = tuple(seg.scale(scale_fac, seg.midpoint) for seg in div_segs)
+            # generate the vertices by 'extruding' along a window height vector.
+            h_vec = base_plane.y * sub_rect_height
+            final_faces = []
+            for seg in scaled_segs:
+                final_faces.append(Face3D(
+                    (seg.p1, seg.p2, seg.p2 + h_vec, seg.p1 + h_vec), base_plane))
+        else:
+            # make a single sub-rectangle at an apporporate sill height
+            max_sill_h = parent_height * 0.99 - (target_area / (parent_base * 0.98))
+            sill_vec = base_plane.y * sill_height if sill_height < max_sill_h \
+                else base_plane.y * max_sill_h
+            seg_init = bottom_seg.move(sill_vec)
+            seg = seg_init.scale(0.98, seg_init.midpoint)
+            h_vec = base_plane.y * (target_area / (parent_base * 0.98))
+            final_faces = [Face3D((seg.p1, seg.p2, seg.p2 + h_vec, seg.p1 + h_vec),
+                                  base_plane)]
+        return final_faces
+
+    def _check_vertices_input(self, vertices):
+        assert isinstance(vertices, (list, tuple)), \
+            'vertices should be a list or tuple. Got {}'.format(type(vertices))
+        assert len(vertices) >= 3, 'There must be at least 3 vertices for a Face3D.' \
+            ' Got {}'.format(len(vertices))
+        assert isinstance(vertices[0], (Point3D, Point3DImmutable)), \
```

Comment:
We can probably remove the first assert and test this for all the vertices to be from type `Point3D` or `Point3DImmutable`. Then we can check the length and that will be all we need to check. Only checking the first vertices doesn't look right to me.

Also as a random comment until I get a chance to do the real review do we really need the Point3D to be `mutable`? I *feel* the default Point3D should be a mutable object and in case one needs an updated point you just create a new one.

Reply:
This is a fair point.  I put back a check for all of the vertices and I realize that this will not be much overhead.  I will say that the main reason why I put it there is to make sure the user used Point3D objects instead of Point2D objects (which will probably be caught on the first vertex) but it is nice to rest assured knowing all vertices meet the criteria.

Now that you mention the idea of all geometry objects being immutable, I really like it.  I originally thought that there would be a use for the mutable points and vectors but, now that I see all of the higher order objects like faces and meshes have to be immutable to avoid crazy amounts of code, we might as well make all points, vectors, lines, rays, and lines immutable as well.  This will simplify the code a lot.  If you agree that this is a good idea, let me know and I will implement it.

Reply:
I would say so. I think it will be much cleaner to do it like that. It is also one of those changes that we should do sooner than later.

Reply:
This is done. The code's a lot cleaner for this.  Thanks for the suggestion.